adoc-mode.el

```diff
@@ -1991,6 +1991,20 @@ LANG is a string, and the returned major mode is a symbol."
          (intern (concat lang "-mode"))
          (intern (concat (downcase lang) "-mode")))))
 
+(defun adoc-map-intervals (fun property &optional beg end object)
+  "Apply FUN to all intervals of PROPERTY in OBJECT in the region from BEG to END.
+FUN is called with two arguments: the beginning and the end of the interval."
+  (unless object (setq object (current-buffer)))
+  (unless beg (setq beg (point-min)))
+  (unless end (setq end (point-max)))
+  (let (end-interval)
+    (while
+	(progn
+	  (setq end-interval (next-single-property-change beg property object end))
+	  (funcall fun beg end-interval)
+	  (setq beg end-interval)
+	  (null (= end-interval end))))))
```

Comment:
Why is this `null` check needed?

Reply:
If `end` is given as buffer position `next-single-property-change` returns that buffer position if no further property change is found. And that also indicates the last interval.